### PR TITLE
SNOW-966444: Fix date binding with qmark

### PR DESCRIPTION
**Pull request by sfc-gh-aling**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes SNOW-966444

the previous fix is incomplete, when TEXT type, date format is not allowed to set.
we switch back to the old milliseconds approach, and handle case according to the spec

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am modifying authorization mechanisms
   - [ ] I am adding new credentials
   - [ ] I am modifying OCSP code
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

File: test/integ/test_bindings.py [hunk 1]
```diff
@@ -480,6 +480,22 @@ def test_binding_bulk_insert_date(conn_cnx, db_parameters):
             )
 
 
+@pytest.mark.skipolddriver
+def test_binding_insert_date(conn_cnx, db_parameters):
+    import logging
+
+    logging.basicConfig(level=logging.DEBUG)
```
Comment (sfc-gh-mkeller) [inline test/integ/test_bindings.py:487]:
```suggestion
```

File: test/integ/test_bindings.py [hunk 2]
```diff
@@ -480,6 +480,22 @@ def test_binding_bulk_insert_date(conn_cnx, db_parameters):
...
+    bind_query = """SELECT TRY_TO_DATE(TO_CHAR(?,?),?)"""
```
Comment (sfc-gh-mkeller) [inline test/integ/test_bindings.py:488]:
```suggestion
    bind_query = "SELECT TRY_TO_DATE(TO_CHAR(?,?),?)"
```

File: src/snowflake/connector/converter.py [hunk 3]
```diff
@@ -366,8 +378,22 @@ def _nonetype_to_snowflake_bindings(self, *_) -> None:
...
+        if value >= date(1970, 1, 1):
+            # according to https://docs.snowflake.com/en/sql-reference/functions/to_date
+            # through test, value in seconds will lead to wrong date
+            # millisecond and nanoseconds are goods
+            milliseconds = _convert_date_to_epoch_milliseconds(value)
+            # if the milliseconds is beyond the range of 31536000000000, we swtich to use nanoseconds
+            # otherwise we will hit overflow error in snowflake
+            return (
+                milliseconds
+                if int(milliseconds) < 31536000000000
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/converter.py:390]: Split this into a full blown if-else clauses please!

**Review by sfc-gh-mkeller — APPROVED** [review]
:ship: :ship:

File: src/snowflake/connector/converter.py [hunk 4]
```diff
@@ -366,8 +378,15 @@ def _nonetype_to_snowflake_bindings(self, *_) -> None:
         return None
 
     def _date_to_snowflake_bindings(self, _, value: date) -> str:
-        # we are binding "TEXT" value for DATE, check function _adjust_bind_type
-        return value.isoformat()
+        milliseconds = _convert_date_to_epoch_milliseconds(value)
+        # according to https://docs.snowflake.com/en/sql-reference/functions/to_date
```
Comment (sfc-gh-aling) [inline src/snowflake/connector/converter.py:382]: create a JIRA

**Review by sfc-gh-stan — APPROVED** [review]

**Pull request merged**